### ISSUE TITLE
PostgreSQL: Json serialization doesn't work with IEnumerable and derivatives

### ISSUE THREAD

**Issue by Hinidu** [jonwagner/Insight.Database#210]
Hi! Thank you again for the great library!
I found an issue with the brand new postgresql json support which can be reproduced by this example:
```csharp
using System.Collections.Generic;
using Insight.Database;
using Npgsql;

namespace Insight
{
    public class Foo
    {
        [Column("id")]
        public int Id { get; set; }

        [Column("list", SerializationMode = SerializationMode.Json)]
        public IEnumerable<int> List { get; set; }
    }

    public class Program
    {
        public static void Main()
        {
            var connection = new NpgsqlConnection(CONNECTION_STRING);
            connection.ExecuteSql("DROP TABLE IF EXISTS foo");
            connection.ExecuteSql("DROP FUNCTION IF EXISTS insert_foo(integer, jsonb)");
            connection.ExecuteSql("CREATE TABLE foo (id integer, list jsonb)");
            connection.ExecuteSql(
                "CREATE FUNCTION insert_foo(id integer, list jsonb) RETURNS void " +
                "AS $body$ INSERT INTO foo VALUES (id, list) $body$ LANGUAGE SQL");
            connection.Execute("insert_foo", new Foo {Id = 1, List = new[] {1, 2, 3, 4, 5}});
        }
    }
}
```
The last line in ```Main``` throws ```NotImplementedException``` with that stack trace:
```
   at Insight.Database.Providers.InsightDbProvider.SetupTableValuedParameter(IDbCommand command, IDataParameter parameter, IEnumerable list, Type listType) in c:\projects.net\Insight\Insight.Database\Insight.Database\Providers\InsightDbProvider.cs:line 288
   at Insight.Database.CodeGenerator.DbParameterGenerator.ListParameterHelper.ConvertListParameterByClass(IDataParameter parameter, IEnumerable list, IDbCommand command, Type listType) in c:\projects.net\Insight\Insight.Database\Insight.Database\CodeGenerator\DbParameterGenerator.cs:line 729
   at Insight.Database.CodeGenerator.DbParameterGenerator.ListParameterHelper.ConvertListParameter(IDataParameter parameter, Object value, IDbCommand command) in c:\projects.net\Insight\Insight.Database\Insight.Database\CodeGenerator\DbParameterGenerator.cs:line 654
   at CreateInputParameters-f27e0ddf-aff7-4c6d-8a94-b609b6445b88(IDbCommand , Object )
   at Insight.Database.DBCommandExtensions.AddParameters(IDbCommand cmd, Object parameters) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBCommandExtensions.cs:line 33
   at Insight.Database.DBConnectionExtensions.CreateCommand(IDbConnection connection, String sql, Object parameters, CommandType commandType, Nullable`1 commandTimeout, IDbTransaction transaction) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBConnectionExtensions.cs:line 276
   at Insight.Database.DBConnectionExtensions.<>c__DisplayClass61.<Execute>b__5f(IDbCommand _, IDataReader __) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBConnectionExtensions.cs:line 328
   at Insight.Database.DBConnectionExtensions.ExecuteAndAutoClose[T](IDbConnection connection, Func`2 getCommand, Func`3 translate, CommandBehavior commandBehavior) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBConnectionExtensions.cs:line 1461
   at Insight.Database.DBConnectionExtensions.ExecuteAndAutoClose[T](IDbConnection connection, Func`2 getCommand, Func`3 translate, Boolean closeConnection) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBConnectionExtensions.cs:line 1429
   at Insight.Database.DBConnectionExtensions.Execute(IDbConnection connection, String sql, Object parameters, CommandType commandType, Boolean closeConnection, Nullable`1 commandTimeout, IDbTransaction transaction, Object outputParameters) in c:\projects.net\Insight\Insight.Database\Insight.Database\Extensions\DBConnectionExtensions.cs:line 324
   at Insight.Program.Main() in z:\Code\Something\Insight\Program.cs:line 91
   at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
   at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
   at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```
I've tried to find the cause myself and ended up with the opinion that Insight just unconditionally handles ```IEnumerable``` always the same way regardless of specified ```SerializationMode``` [here](https://github.com/jonwagner/Insight.Database/blob/master/Insight.Database/CodeGenerator/DbParameterGenerator.cs#L183-L196).

PS. I would like to work on this issue myself but I didn't manage to build and run Insight yet: I can't install psake on my work machine because it needs administrator privileges and I can't run it on my personal linux box because PowerShell is Windows specific. Also I constantly have issues with running your Vagrantfile on any machine - it just can't establish the connection to the guest OS. Though I hope I will be able to solve some of these problems and maybe produce a PR.

Thank you in advance!

**Comment by Hinidu** [comment]
The message in the exception is ```Cannot set up the table valued parameter for parameter list.. Have you loaded the provider that suppoerts NpgsqlCommand?```.

**Comment by jonwagner** [comment]
You're probably right about the enumerable issue. Let's see if it's an east fix.

**Comment by jonwagner** [comment]
Looks like just a small fix in the json serializer. This will be fixed in v5.2.6

**Comment by Hinidu** [comment]
Thank you! It works for me now.
